### PR TITLE
[experimental/batched_rpa] separate metadata into own kernel, alias q_hbm/o_hbm, int16 support, add flags

### PR DESCRIPTION
**Pull request by a1yssan13**
added the following changes: 
- Porting the int16 accumulation optimization from latest RPA3 commit
- Configurable flags `mask_v` to apply mask to v tensor (otherwise would apply to QK). fuse_accum allows the accumulation logic to not be in a conditional branch 
- Separating out the schedule metadata into a separate kernel so that we can cache the same schedule across multiple layers in the model. 
- Ported strided_load from RPA3

### CODE REVIEW

File: tpu_inference/kernels/experimental/batched_rpa/flash_attention.py [hunk 1]
```diff
@@ -14,19 +14,17 @@
 
 import jax.numpy as jnp
 from jax import lax
+from jax.experimental import pallas as pl
+from jax.experimental.pallas import tpu as pltpu
 
 from tpu_inference.kernels.experimental.batched_rpa import \
     schedule as rpa_schedule
 
 _NUM_LANES = 128
```
Comment (kyuyeunk) [inline tpu_inference/kernels/experimental/batched_rpa/flash_attention.py]: Is this needed?

File: tpu_inference/kernels/experimental/batched_rpa/flash_attention.py [hunk 2]
```diff
@@ -83,22 +81,43 @@ def flash_attention(
...
+    for b_idx in range(config.batch_size):
+        kv_idx_b = (lax.broadcasted_iota(int_ty, (k_heads, tq, s), 2) +
+                    processed_kv_len[b_idx])
+        q_idx_b = (lax.broadcasted_iota(jnp.int32, (k_heads, tq, s), 1) //
+                   config.num_q_heads_per_kv_head
+                   ).astype(int_ty) + processed_q_len[b_idx]
+
+        eff_kv_len_b = effective_kv_len[b_idx]
+        mask_b = q_idx_b < eff_kv_len_b
+        mask_b &= q_idx_b >= kv_idx_b
```
Comment (kyuyeunk) [inline tpu_inference/kernels/experimental/batched_rpa/flash_attention.py]: Nit: recommend replacing '&' with something more explicit like jnp logical and. Similar comments for all the other usages

File: tpu_inference/kernels/experimental/batched_rpa/flash_attention.py [hunk 3]
```diff
@@ -83,22 +81,43 @@ def flash_attention(
...
-        qk_seq_masks.append(mask_i)
-    mask = jnp.stack(qk_seq_masks, axis=0)
+            mask_b &= q_idx_b < kv_idx_b + config.sliding_window
+
+        if not config.mask_v:
+            mask_b &= kv_idx_b < eff_kv_len_b
+
+        qk_masked.append(jnp.where(mask_b, qk[b_idx], config.mask_value))
+
+        if config.mask_v:
```
Comment (kyuyeunk) [inline tpu_inference/kernels/experimental/batched_rpa/flash_attention.py:91]: I think this if else can be merged with if statement above?

File: tpu_inference/kernels/experimental/batched_rpa/kernel.py [hunk 4]
```diff
@@ -92,10 +90,10 @@ def rpa_body(
             def strided_load(ref, start, sz, step, *, dtype=None):
                 assert schedule_lib.get_dtype_packing(ref.dtype) == 1
                 assert len(ref.shape) == 2
-                num_rows, num_cols = ref.shape  # noqa
-                assert num_cols % 128 == 0
-                folds = num_cols // 128
-                ref = ref.reshape(num_rows * folds, 128)
```
Comment (kyuyeunk) [inline tpu_inference/kernels/experimental/batched_rpa/kernel.py:98]: Let's roll back this change since the previous version is easier to read.

File: tpu_inference/kernels/experimental/batched_rpa/kernel.py [hunk 5]
```diff
@@ -191,6 +189,10 @@ def _convert_to_target_bitwidth(val, target_bitwidth: int):
             processed_q_len = []
             processed_kv_len = []
             effective_kv_len = []
+            int_ty = jnp.int32
+            if (schedule_lib.get_dtype_packing(config.q_dtype) != 1
```
Comment (kyuyeunk) [inline tpu_inference/kernels/experimental/batched_rpa/kernel.py]: This logic seems to be used multiple times. Maybe create a separate function?

File: tpu_inference/kernels/experimental/batched_rpa/schedule.py [hunk 6]
```diff
@@ -81,9 +81,12 @@ class RPAConfig:
     q_scale: Any = None
     k_scale: Any = None
     v_scale: Any = None
-    vmem_limit_bytes: int | None = None
+    vmem_limit_bytes: int | None = 60 * 1024 * 1024
```
Comment (kyuyeunk) [inline tpu_inference/kernels/experimental/batched_rpa/schedule.py]: Revert this change? I think we already set default value using get tpu info?

File: tpu_inference/kernels/experimental/batched_rpa/schedule.py [hunk 7]
```diff
@@ -240,6 +243,15 @@ def out_shape(cls, config: RPAConfig):
...
+    def hbm_specs(cls, config: RPAConfig):
+        """Returns a Pytree of BlockSpecs in HBM matching the output structure."""
+
+        def wrapper(shape, is_smem=False):
+            return pl.BlockSpec(memory_space=pltpu.HBM, block_shape=shape)
+
+        return cls._map_shapes(config, wrapper)
+
     @classmethod
     def test_specs(cls, config: RPAConfig):
```
Comment (kyuyeunk) [inline tpu_inference/kernels/experimental/batched_rpa/schedule.py]: What is test specs?

File: tpu_inference/kernels/experimental/batched_rpa/schedule.py [hunk 8]
```diff
@@ -267,15 +279,35 @@ def smem_specs(cls, config: RPAConfig):
...
     schedule: RPASchedule,
     lane_lengths_ref,
+    dma_sem,
     *,
     config: RPAConfig,
 ):
-    """Generates metadata for RPA scheduling."""
+    """Generates the HBM-to-VMEM DMA schedule
+
+  This kernel:
```
Comment (kyuyeunk) [inline tpu_inference/kernels/experimental/batched_rpa/schedule.py:304]: Fix indentation to 4 spaces

File: tpu_inference/kernels/experimental/batched_rpa/schedule.py [hunk 9]
```diff
@@ -311,6 +343,11 @@ def q_loop(q_idx, _):
             end_k_idx = jnp.minimum(n_k, end_k_idx_causal)
 
             def k_loop(k_idx, curr_ptr):
+                # pl.debug_check(
```
Comment (kyuyeunk) [inline tpu_inference/kernels/experimental/batched_rpa/schedule.py]: Remove?

File: tpu_inference/kernels/experimental/batched_rpa/flash_attention.py (hunk 3 above)
Comment (a1yssan13) [reply tpu_inference/kernels/experimental/batched_rpa/flash_attention.py:91]: Hmm i think 

> qk_masked.append(jnp.where(mask_b, qk[b_idx], config.mask_value))

needs to go in between? which is why i separated it out
